### ISSUE TITLE
When new Value Object available inside an entity is created, it is being treated as reference change inside of Value change

### ISSUE THREAD

**Issue by ghost** [javers/javers#732]
Considering a case where my entity has few Value Objects but they are not having any values, so are null.
When these value objects are created and added to the entity object, this is being considered as reference change, wherein I think it should be taken as value change.

**Comment by bartoszwalacik** [comment]
please provide a test case which will clarify your expectations

**Comment by vandanagoyal-ak** [comment]
Considering an entity and a Value object inside this entity.

@Entity
class Employee{
private String id;
private String name;

@Embedded
private Address address;
}

@ValueObject
@Embeddable
class Address{
private String city;
private String street;
private String country;
}

I compare two employee objects.

1) Employee empOld = Employee.builder().id("1234").build();
2)Employee empNew = Employee.builder.id("1234).name("xyz").address(new Address("Berlin","xyz street", "Germany")).build();

Now, 

Javers javers = JaversBuilder
                    .javers()
                    .withListCompareAlgorithm(LEVENSHTEIN_DISTANCE)
                    .build();
Diff diff = javers.compare(empOld, empNew);
List<ReferenceChange> referenceChanges = diff.getChangesByType(ReferenceChange.class);

Addition of new 'ValueObject' Address in employee entity here creates a change of type 'ReferenceChange'. 

It also creates a 'NewObject' Change - that is fine.
But creating a 'ReferenceChange' is I think wrong because Address is a value object and not an entity or a reference type. I expect that addition of Address should also be taken as 'ValueChange' here.

Also, 
if empOld contained an address already and then compared with empNew, then it is taken as 'ValueChange' which is correct but for CREATE cases(when valueObject is newly added to the entity), it behaves differently and creates a 'ReferenceChange'.

The behavior for ValueObjects should be same irrespective of the fact that they are updated or newly added as they are not Reference objects and should not be part of "ReferenceChange".

I hope this use case clarifies. Thank you.

**Comment by vandanagoyal-ak** [comment]
Hi, did my use case and explanation clarified the issue?

**Comment by bartoszwalacik** [comment]
Yeah, I will try to manage it during this weekend. This kind of change in javers-core would be a breaking change but fortunately, javers 4 release is coming ...

**Comment by bartoszwalacik** [comment]
So you expect sth like this: ?

```groovy
    class Employee{
        @Id String id
        Address address
    }

    class Address{
        private String city
        private String street
    }

    def "should generate ValueChange and not ReferenceChange when null is changed to ValueObject"() {
        given:
        def javers = JaversBuilder.javers().build()

        when:
        def diff = javers.compare(new Employee(id: "1"),
                                  new Employee(id: "1", address:new Address(city: "Berlin", street: "x")))

        diff.changes.forEach{println it.toString() + " at:" + it.affectedGlobalId}

        then:
        diff.changes.size() == 2
        
        diff.getChangesByType(NewObject).size() == 1
        def valueChanges = diff.getChangesByType(ValueChange)
        valueChanges.size() == 1
        valueChanges[0].left == null
        valueChanges[0].right instanceof ValueObjectId
        valueChanges[0].right.value().endsWith('Case$Employee/1#address')
    }
```

**Comment by bartoszwalacik** [comment]
After rethinking, in my opinion changing `ReferenceChange` to `ValueChange` in this case would be wrong because it would be inconsistent with the current JaVers' diff and persistence model.
As a diff result, you get null on the left and `ValueObjectId` on the right, maybe it's not very useful, but it's consistent. `ValueObjectId` is not a Value, it's a reference.

What I can agree is:
* adding a flag to disable generating `ReferenceChange` when null is changed to `ValueObject`
* or dividing `ReferenceChange` into two classes: `InstanceIdChange` and `ValueObjectIdChange`, both would extend abstract `ReferenceChange`. This option looks better for me.

**Comment by vandanagoyal-ak** [comment]
* In your first option of adding a flag to disable generating `ReferenceChange` when null is changed to `ValueObject`, in this case, then will it generate `ValueChange`? With `ValueObjectId` on the right, will I be anyhow able to fetch the new values of the value object?

* For the second option, it is fine but there will still be inconsistency in the fact that when `oldValueObject` is changed to `newValueObject` (update case) it creates a `ValueChange`. But when null is changed to `ValueObject`, it is a `ReferenceChange` (create case).

**Comment by bartoszwalacik** [comment]
* yes

* There is no inconsistency here. `ReferenceChange` would be abstract class, concrete class would be either `ValueObjectIdChange` or `InstanceIdChange`.

**Comment by bartoszwalacik** [comment]
I think that null Value Object should be treated as empty Value Object. So this test

```groovy
class ValueObjectRefChangeCase extends Specification {

    class Employee{
        @Id String id
        Address address
    }

    class Address{
        private String city
        private String street
    }

    def "should generate ValueChange and not ReferenceChange when null is changed to ValueObject"() {
        given:
        def javers = JaversBuilder.javers().build()

        when:
        def diff = javers.compare(new Employee(id: "1", address:null),
                                  new Employee(id: "1", address:new Address(city: "Berlin", street: "x")))

        println diff.prettyPrint()

        then:
        diff.changes.size() == 2

        diff.changes.forEach {
            assert it instanceof ValueChange
            assert it.left == null
            assert it.right != null
        }
    }
}
```

should calculate:

```
Diff:
* changes on org.javers.core.cases.ValueObjectRefChangeCase$Employee/1 :
  - 'address.city' value changed from '' to 'Berlin'
  - 'address.street' value changed from '' to 'x'
```

now, we have not very useful diff:

```
Diff:
* new object: org.javers.core.cases.ValueObjectRefChangeCase$Employee/1#address
* changes on org.javers.core.cases.ValueObjectRefChangeCase$Employee/1 :
  - 'address' reference changed from '' to '...ValueObjectRefChangeCase$Employee/1#address'
```

**Comment by divikrckr** [comment]
def diff = javers.compare(new Employee(id: "1",state:null),
                                  new Employee(id: "1", state:"Delhi"))
Will it be a value change or ObjectAdded where state is of string type??

**Comment by jurna** [comment]
What about progress of this issue? Currently it is not possible to get diff of what value object was added to an antity (you can see only the reference, but not the value itself) . In this case diff is not very usefull, it does not show the actual values that were added. It is a major issue for us when calculating diff of two different objects

**Comment by bartoszwalacik** [comment]
@jurna this is open source project, you want a feature -- contribute

**Comment by bartoszwalacik** [comment]
fixed in 6.0.0-rc1

**Comment by jyothi462** [comment]
Thanks for this fix. This definitely is helpful to check diff in a better way. 
I just wanted to understand how to differentiate between value changes of nested fields with same names

```
    class Employee{
        @Id String id
        Address homeAddress
        Address officeAddress
    }

    class Address{
        private String city
        private String street
    }
    
    def diff = javers.compare(new Employee(id: "1", address:null),
                                  new Employee(id: "1", address:new Address(city: "Berlin", street: "x"), address:new Address(city: "Berlin", street: "y")))
```

In this example, the resulting propertyNames would be 'city' and 'street' for both homeAddress and officeAddress.
I think it would be better if we can have something like 'homeAddress.city',  'homeAddress.street',  'officeAddress.city' and  'officeAddress.street'
